core/src/main/java/org/elasticsearch/action/bulk/MappingUpdatePerformer.java

```diff
@@ -20,48 +20,25 @@
 package org.elasticsearch.action.bulk;
 
 import org.elasticsearch.action.index.IndexRequest;
-import org.elasticsearch.common.Nullable;
 import org.elasticsearch.index.engine.Engine;
 import org.elasticsearch.index.shard.IndexShard;
 
 import java.util.Objects;
 
 public interface MappingUpdatePerformer {
```

Comment:
Currently `MappingUpdatePerformer#updateMappingsIfNeeded` prepares index operations for primary, besides updating the mappings on master and `MappingUpdatePerformer#verifyMappings` repeats the same index operation preparation. IMO, we could not duplicate the same index preparation operation and it would be cleaner if the interface was along these lines:

```java
    interface MappingUpdatePerformer {

        /**
         * Determine if any mappings need to be updated before executing <code>operation</code>, 
         * and update mapping for document <code>type</code> in <code>shardId</code> on the master node if necessary. 
         * Throws a failure Exception in the event updating the mappings fails
         */
        void updateMappingsIfNeeded(final Engine.Index operation,
                                    final ShardId shardId,
                                    final String type) throws Exception;

        /**
         * Throws a {@code ReplicationOperation.RetryOnPrimaryException} if <code>operation</code> needs to be
         * retried on the primary <code>shardId</code> due to the mappings not being present yet
         */
        void verifyMappings(final Engine.Index operation,
                            final ShardId shardId) throws Exception;
    }
```
This means `TransportShardBulkAction#prepareIndexOperationOnPrimary` should be called in `TransportShardBulkAction#executeIndexRequestOnPrimary` before calling `MappingUpdatePerformer#updateMappingsIfNeeded`